### PR TITLE
AP_Beacon and EKF2 enhancements for Non-GPS Loiter

### PR DESCRIPTION
**Pull request by rmackay9**
This set of commits includes:

- AP_Beacon library which accepts messages from an ArduinoUno/Pozyx which provides distances from beacons placed as described on this wiki page: http://ardupilot.org/copter/docs/common-pozyx.html
- EKF2 enhancements to use the beacon distances for local position estimation without using a GPS
- SITL simulator for AP_Beacon
- some unrelated changes to resolve some compiler warnings

### CODE REVIEW

**Comment by OXINARF** [comment]
Good job! I've only had a quick look but it seemed good (except that it looks like EKF lost the ability to go from an absolute aid to a relative aid). I was a little surprised that this was done as an absolute aid instead of a relative one as the optical flow.
As it been tested what influence it has if the system transitions from no-GPS (with beacons providing the position) to getting GPS reception (with beacons or not)? Does the position of the origin have to be precise if used together with GPS?

A couple of questions:
- is the configuration of 4 anchors in a specific rectangle a limitation of the EKF or the script running in the Arduino? If it is the EKF, is there any plan to remove this in the future?
- does this also provide height? I think Randy said in a dev call it didn't, but from my quick look at code it looks like it does.

It's awesome that ArduPilot is getting this!

**Comment by rmackay9** [comment]
Thanks!

The requirement for 4 anchors comes partly from the Pozyx system.  It requires at least 3.  I made it 4 because at least at first, the range was quite short so I thought one at each corner would add some redundancy.  The range has improved since then so we could reduce it to 3 if people really want.

The pozyx can provide height but it's not very good.  The EKF actually only uses the position from the Pozyx at startup and during this phase we assume the vehicle is on the ground so the height isn't too important.

I'll leave Paul to answer the other more EKF-ish questions.

**Comment by rmackay9** [comment]
With @peterbarker's help I've added some extra checks to the Log_Write_EKF2 function but it seems that somehow the ahrs.get_NavEKF2().getRangeBeaconDebug() function can return uninitailised values.  If @priseborough can figure out why then we should be able to remove these extra checks I've added.

**Comment by OXINARF** [comment]
I was actually wondering if more anchors could be used and specially if they could be placed in any way we wanted. Think of a maze full of anchors - as long as you previously configure their location, could we get location as we go through the maze?

**Comment by muramura** [comment]
In casting, it is better to use static_cast etc.
At compile time, if type conversion is inappropriate, an error is detected.
This improves the soundness of the value.

**Comment by muramura** [comment]
I think.
The structure uses packed.
If you can, do not use it.
Also, 
if you use it,It is better to arrange variables that conscious of access time.

**Comment by muramura** [comment]
Editor settings are written in the file header.
I think that it would be better not to write in editor settings
ex:
// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-

**Comment by muramura** [comment]
In the case of a character string ending with "\n" without using a parser in the printf statement, I think that the println statement is better.

File: libraries/AP_Beacon/AP_Beacon_Pozyx.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+{
+    // check crc
+    uint8_t checksum = 0;
+    checksum ^= parse_msg_id;
+    checksum ^= parse_msg_len;
+    for (uint8_t i=0; i<linebuf_len-1; i++) {
+        checksum ^= linebuf[i];
+    }
+    // return if failed checksum check
+    if (checksum != linebuf[linebuf_len-1]) {
```
Comment (muramura) [inline libraries/AP_Beacon/AP_Beacon_Pozyx.cpp]: When exclusive OR is performed up to the calculation result, it becomes 0.
Therefore,
I think that it should be judged by whether the value of cecksum is 0 or not.

File: libraries/AP_Beacon/AP_Beacon_SITL.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        return;
+    }
+
+    uint8_t beacon_id = next_beacon;
+    next_beacon = (next_beacon+1) % NUM_BEACONS;
+
+    // truth location of the flight vehicle
+    Location current_loc;
+    current_loc.lat = sitl->state.latitude * 1.0e7;
+    current_loc.lng = sitl->state.longitude * 1.0e7;
```
Comment (muramura) [inline libraries/AP_Beacon/AP_Beacon_SITL.cpp]: The longitude here is "lng".
It is better to unify the words.

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 3]
```diff
@@ -648,4 +648,113 @@ void NavEKF2_core::readAirSpdData()
...
+        numRngBcnsChecked++;
+
+        // move to next beacon, wrap index if necessary
+        index++;
+        if (index >= N_beacons) {
+            index = 0;
+        }
+
+        // check that the beacon is healthy and has new data
+        if (beacon != nullptr &&
```
Comment (muramura) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp]: L662 is checking the value of beacon.
I think this check is unnecessary.

File: libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp [hunk 4]
```diff
@@ -648,4 +648,113 @@ void NavEKF2_core::readAirSpdData()
...
+            // indicate we have new data to push to the buffer
+            newDataToPush = true;
+
+            // update the last checked index
+            lastRngBcnChecked = index;
+        }
+    }
+
+    // Check if the beacon system has returned a 3D fix
+    if (beacon != nullptr && beacon->get_vehicle_position_ned(beaconVehiclePosNED, beaconVehiclePosErr)) {
```
Comment (muramura) [inline libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp]: L662 is checking the value of beacon.
I think this check is unnecessary.

File: libraries/AP_NavEKF2/AP_NavEKF2_RngBcnFusion.cpp [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+        Kfusion[14] = -t26*(P[14][6]*t4*t9+P[14][7]*t3*t9+P[14][8]*t2*t9);
+        if (!inhibitMagStates) {
+            Kfusion[16] = -t26*(P[16][6]*t4*t9+P[16][7]*t3*t9+P[16][8]*t2*t9);
+            Kfusion[17] = -t26*(P[17][6]*t4*t9+P[17][7]*t3*t9+P[17][8]*t2*t9);
+            Kfusion[18] = -t26*(P[18][6]*t4*t9+P[18][7]*t3*t9+P[18][8]*t2*t9);
+            Kfusion[19] = -t26*(P[19][6]*t4*t9+P[19][7]*t3*t9+P[19][8]*t2*t9);
+            Kfusion[20] = -t26*(P[20][6]*t4*t9+P[20][7]*t3*t9+P[20][8]*t2*t9);
+            Kfusion[21] = -t26*(P[21][6]*t4*t9+P[21][7]*t3*t9+P[21][8]*t2*t9);
+        } else {
+            for (uint8_t i=16; i<=21; i++) {
```
Comment (muramura) [inline libraries/AP_NavEKF2/AP_NavEKF2_RngBcnFusion.cpp]: I think.
Used memset is fast.

**Comment by rmackay9** [comment]
@OXINARF, yes, I think as a next step we could add support for more anchors.  The EKF currently supports up to 10 but I think even that is a bit arbitrary.  I think the only real limitation is that it takes about 35ms for the pozyx on the vehicle to get the distance from each beacon.  If we want to maintain a 7hz ~ 10hz update rate then we need to add some logic to only ask for beacons that are likely to be close to us (or ask for the close ones at a high rate, farther ones and a lower rate).

I've been thinking that once we have the first few beacon placed, we could use them to find the location of other beacons.  As long as we tell the EKF where the beacons are then it can consume them appropriately.  They don't need to all be a square actually... it's actually just the first 2 that must be at 90degrees from the origin.

**Comment by priseborough** [comment]
@OXINARF in answer to your comments:

1) Relative aiding mode has been set up for optical flow and other methods where no position measurement is available. Using it with beacons would have required changes to that mode that could have broken optical flow.

2) It has not been tested in conjunction with GPS. Using beacons in conjunction with GPS will require modification of the calculation to compute the beacon constellation offset from 1D to 3D so that the beacon measurements do not 'flight' with the GPS. This is the same reason we do not fuse  in measurements from two GPS receivers. Even with a 'precise' origin for the beacons, the GPS solution can easily drift by 5 metres during a flight.

File: libraries/AP_NavEKF2/AP_NavEKF2_RngBcnFusion.cpp (hunk 5 above)
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_RngBcnFusion.cpp]: How would I memset a specific range of data in an array?

Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_RngBcnFusion.cpp]: Would the following work?
memset(&Kfusion[16], 0, 24)

**Comment by priseborough** [comment]
I will rebase this branch on latest master, include my review changes and push to here: https://github.com/priseborough/ardupilot/tree/prise-beacon9

**Comment by rmackay9** [comment]
@priseborough, ok, great thanks, I'll pick up any changes from your prise-beacon9 branch.  For me, the major hold-up on pushing this to master is just the initialisation issue in the Log_Write_EKF2 function.  All the isnan() checks were required because that function can be called before any of the variables we are logging are initialised.

**Comment by priseborough** [comment]
I have pushed the rebase and all review changes to https://github.com/priseborough/ardupilot/tree/prise-beacon9

**Comment by rmackay9** [comment]
I've fixed the other two issues raised by @muramura so closing this now.  Thanks!

**Pull request closed without merging**